File: src/main/java/toker/warbandscripts/panel/service/BanService.java
package toker.warbandscripts.panel.service;

import org.springframework.data.crossstore.ChangeSetPersister;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import toker.warbandscripts.panel.bean.SelectedServerId;
import toker.warbandscripts.panel.entity.Ban;
import toker.warbandscripts.panel.entity.Player;
import toker.warbandscripts.panel.entity.Server;
import toker.warbandscripts.panel.repository.BanRepository;
import toker.warbandscripts.panel.repository.PanelUserRepository;
import toker.warbandscripts.panel.repository.PlayerRepository;
import toker.warbandscripts.panel.repository.ServerRepository;

import java.io.*;
import java.sql.Timestamp;
import java.time.Instant;
import java.util.Calendar;
import java.util.List;

@Service
public class BanService {

    private PanelConfigurationService configurationService;

    private BanRepository banRepository;
    private PlayerRepository playerRepository;
    private PanelUserRepository panelUserRepository;
    private ServerRepository serverRepository;

    public BanService(PanelConfigurationService configurationService,
                      BanRepository banRepository,
                      PlayerRepository playerRepository,
                      PanelUserRepository panelUserRepository,
                      ServerRepository serverRepository) {
        this.configurationService = configurationService;
        this.banRepository = banRepository;
        this.playerRepository = playerRepository;
        this.panelUserRepository = panelUserRepository;
        this.serverRepository = serverRepository;
    }

    @Scheduled(fixedDelay = 1000 * 30)
    public void refreshBanList() throws IOException {
        for (Server server : serverRepository.findAll()) {
            File serverFolder = new File(new File(server.getExePath()).getParent());
            File banFile = new File(serverFolder, "banlist.txt");
            FileWriter writer = new FileWriter(banFile);

            for (Ban ban : banRepository.findAllByServerId(server.getId())) {
                if (ban.isUndone())
                    continue;

                if (ban.isPermanent()) {
                    writer.write(String.valueOf(ban.getPlayerUniqueId()));
                    writer.write(System.lineSeparator());
                    continue;
                }

                Calendar calendar = Calendar.getInstance();
                calendar.setTime(ban.getTime());
                calendar.add(Calendar.MINUTE, ban.getMinutes());

                if (calendar.toInstant().isAfter(Instant.now())) {
                    writer.write(String.valueOf(ban.getPlayerUniqueId()));
                    writer.write(System.lineSeparator());
                }
            }

            writer.flush();
            writer.close();
        }
    }

    public Ban banPlayer(Ban ban) {
        String adminName = (String)SecurityContextHolder
                .getContext().getAuthentication().getPrincipal();

        ban.setPanelUser(panelUserRepository.findByUsername(adminName));
        ban.setServer(serverRepository.getOne(SelectedServerId.get()));
        ban.setTime(Timestamp.from(Instant.now()));
        ban.setUndone(false);

        return banRepository.saveAndFlush(ban);
    }

    public Ban undoBan(int banId) throws ChangeSetPersister.NotFoundException {
        Ban ban = banRepository.findById(banId)
                .orElseThrow(ChangeSetPersister.NotFoundException::new);
        ban.setUndone(true);
        return banRepository.saveAndFlush(ban);
    }

    public void undoAllBansForPlayer(int playerUniqueId) {
        banRepository.undoAllByPlayerUniqueId(playerUniqueId, SelectedServerId.get());
    }

    public List<Ban> getBansOfPlayer(int playerId) {
        Player player = playerRepository.findById(playerId).orElse(null);
        if (player != null) {
            return banRepository.findAllByPlayerUniqueId(player.getUniqueId());
        }
        return null;
    }

    public List<Ban> getBansOfAdmin(int adminId) {
        return banRepository.findAllByPanelUserId(adminId);
    }
}

File: src/main/java/toker/panel/service/BanService.kt
package toker.panel.service

import org.springframework.context.annotation.Profile
import org.springframework.data.crossstore.ChangeSetPersister
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Service
import toker.panel.bean.SelectedServerId
import toker.panel.entity.Ban
import toker.panel.repository.BanRepository
import toker.panel.repository.PanelUserRepository
import toker.panel.repository.PlayerRepository
import toker.panel.repository.ServerRepository
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.sql.Timestamp
import java.time.Instant
import java.util.*

@Service
class BanService(private val banRepository: BanRepository,
                 private val playerRepository: PlayerRepository,
                 private val panelUserRepository: PanelUserRepository,
                 private val serverRepository: ServerRepository) {

    @Scheduled(fixedDelay = 1000 * 30)
    @Throws(IOException::class)
    fun refreshBanList() {
        for (server in serverRepository.findAll()) {
            val serverFolder = File(File(server.exePath!!).parent)
            val banFile = File(serverFolder, "banlist.txt")
            val writer = FileWriter(banFile)
            for (ban in banRepository.findAllByServerId(server.id!!)) {
                if (ban.isUndone!!) continue
                if (ban.isPermanent!!) {
                    writer.write(ban.playerUniqueId.toString())
                    writer.write(System.lineSeparator())
                    continue
                }
                val calendar = Calendar.getInstance()
                calendar.time = ban.time
                calendar.add(Calendar.MINUTE, ban.minutes!!)
                if (calendar.toInstant().isAfter(Instant.now())) {
                    writer.write(ban.playerUniqueId.toString())
                    writer.write(System.lineSeparator())
                }
            }
            writer.flush()
            writer.close()
        }
    }

    fun banPlayer(ban: Ban): Ban {
        val adminName = SecurityContextHolder
                .getContext().authentication.principal as String
        ban.panelUser = panelUserRepository.findByUsername(adminName)
        ban.server = serverRepository.getOne(SelectedServerId)
        ban.time = Timestamp.from(Instant.now())
        ban.isUndone = false
        return banRepository.saveAndFlush(ban)
    }

    @Throws(ChangeSetPersister.NotFoundException::class)
    fun undoBan(banId: Int): Ban {
        val ban = banRepository.findById(banId)
                .orElseThrow { ChangeSetPersister.NotFoundException() }
        ban.isUndone = true
        return banRepository.saveAndFlush(ban)
    }

    fun undoAllBansForPlayer(playerUniqueId: Int) {
        banRepository.undoAllByPlayerUniqueId(playerUniqueId, SelectedServerId)
    }

    fun getBansOfPlayer(playerId: Int): List<Ban>? {
        val player = playerRepository.findById(playerId).orElse(null)
        return if (player != null) {
            banRepository.findAllByPlayerUniqueId(player.uniqueId!!)
        } else null
    }

    fun getBansOfAdmin(adminId: Int): List<Ban> {
        return banRepository.findAllByPanelUserId(adminId)
    }
}